File: solutions/csharp/spiral-matrix/1/SpiralMatrix.cs
using Position = (int Row, int Column);

public class SpiralMatrix
{
    public static int[,] GetMatrix(int size)
    {
        var matrix = new int[size, size];

        Populate(0, 1, size, matrix);

        return matrix;
    }

    private static void Populate(int offset, int start, int width, int[,] matrix)
    {
        if (width == 0)
        {
            return;
        }

        if (width == 1)
        {
            matrix[offset, offset] = start;
            return;
        }

        var number = start;

        foreach (var coordinate in SquareCoordinates(offset, width))
        {
            matrix[coordinate.Row, coordinate.Column] = number;
            number++;
        }

        Populate(offset + 1, number, width - 2, matrix);
    }

    private static IEnumerable<Position> SquareCoordinates(int offset, int width)
    {
        var end = offset + width;

        for (var column = offset; column < end; column++)
        {
            yield return (offset, column);
        }

        for (var row = offset + 1; row < end; row++)
        {
            yield return (row, end - 1);
        }

        for (var column = end - 2; column >= offset; column--)
        {
            yield return (end - 1, column);
        }

        for (var row = end - 2; row > offset; row--)
        {
            yield return (row, offset);
        }
    }
}

File: solutions/scala/spiral-matrix/1/src/main/scala/SpiralMatrix.scala
object SpiralMatrix:
  private val Directions = Vector((0, 1), (1, 0), (0, -1), (-1, 0))

  def spiralMatrix(size: Int): List[List[Int]] =
    if size == 0 then List.empty
    else
      val matrix = Array.fill(size, size)(0)
      var row = 0
      var column = 0
      var directionIndex = 0

      for number <- 1 to size * size do
        matrix(row)(column) = number

        val (rowDelta, columnDelta) = Directions(directionIndex)
        val nextRow = row + rowDelta
        val nextColumn = column + columnDelta

        if isOutside(nextRow, nextColumn, size) || matrix(nextRow)(nextColumn) != 0 then
          directionIndex = (directionIndex + 1) % Directions.size

        val (newRowDelta, newColumnDelta) = Directions(directionIndex)
        row += newRowDelta
        column += newColumnDelta

      matrix.map(_.toList).toList

  private def isOutside(row: Int, column: Int, size: Int): Boolean =
    row < 0 || row >= size || column < 0 || column >= size
